File: 0539-Minimum Time Difference/0539-Minimum Time Difference.cpp
class Solution {
public:
    int findMinDifference(vector<string>& timePoints) {
        vector<int> bucket(1440);
        for (string& s : timePoints) {
            int index = stoi(s.substr(0, 2)) * 60 + stoi(s.substr(3));
            if (bucket[index]++) {
                return 0;
            }
        }
        
        int first = -1, last = -1;
        int minDiff = INT_MAX;
        for (int i = 0; i < 1440; ++i) {
            if (bucket[i]) {
                if (first == -1) {
                    first = i;
                }
                else {
                    minDiff = min(minDiff, i - last);
                }
                last = i;
            }
        }
        
        return min(minDiff, 1440 - (last - first));
    }
};

File: 0501-0600/0539-Minimum Time Difference/0539-Minimum Time Difference.py
class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        bucket = [0] * 1440
        for time in timePoints:
            index = int(time[:2]) * 60 + int(time[3:])
            bucket[index] += 1
            if bucket[index] > 1:
                return 0
        first = last = -1
        minDiff = 1440
        for i, b in enumerate(bucket):
            if b:
                if first == -1:
                    first = i
                else:
                    minDiff = min(minDiff, i - last)
                last = i
        return min(minDiff, 1440 - (last - first))
